### PR TITLE
User verification, refresh token and password reset functionality (+API documentation)

### PR DESCRIPTION
**Pull request by stefanvanherwijnen**
Please let me know if something has to be changed or improved.

### CODE REVIEW

**Comment by mstaack** [comment]
Thanks for your hard work! Give me some time to review please

File: app/Extensions/Authentication/PasetoAuthGuard.php [hunk 1]
```diff
@@ -167,6 +167,38 @@ public function setUser(Authenticatable $user)
...
+    * @throws Exception
+    * @throws InvalidKeyException
+    * @throws InvalidPurposeException
+    * @throws PasetoException
+    * @throws TypeError
+    */
+    public function refresh(Request $request)
+    {
+        $this->setRequest($request);
+        $refresh_token = $this->getTokenFromRequest();
```
Comment (mstaack) [inline app/Extensions/Authentication/PasetoAuthGuard.php]: please use camelCase

File: app/Extensions/Authentication/PasetoAuthGuard.php [hunk 2]
```diff
@@ -167,6 +167,38 @@ public function setUser(Authenticatable $user)
...
+        try {
+            $this->token = $this->parseToken($refresh_token);
+            $this->user = $this->user();
+            if ($this->user->refresh_token == $refresh_token) {
```
Comment (mstaack) [inline app/Extensions/Authentication/PasetoAuthGuard.php]: always use `===`

File: app/Extensions/Authentication/PasetoAuthGuard.php [hunk 3]
```diff
@@ -167,6 +167,38 @@ public function setUser(Authenticatable $user)
...
+                return $this->generateTokenForUser();
+            }
+            else {
+                return false;
+            }
+        } catch (PasetoException $e) {
+            return false;
+        }
+
```
Comment (mstaack) [inline app/Extensions/Authentication/PasetoAuthGuard.php]: remove whitespaces

File: app/Http/Controllers/AuthController.php [hunk 4]
```diff
@@ -52,26 +76,134 @@ public function login(Request $request)
...
+        if(isset($verifyUser) ){
+            $verifyUser->user->verified = 1;
+            $verifyUser->user->save();
+            $verifyUser->delete();
+            return response()->json([
+                'status' => 'success',
+                'message' => 'verified'
+            ]);
+        }
+        else {
```
Comment (mstaack) [inline app/Http/Controllers/AuthController.php]: no else needed

File: app/Http/Controllers/AuthController.php [hunk 5]
```diff
@@ -52,26 +76,134 @@ public function login(Request $request)
...
+
+    /**
+    * @param String $token
+    *
+    * @return Response
+    */
+    public function verify($token)
+    {
+        $verifyUser = VerifyUser::where('token', $token)->first();
+        if(isset($verifyUser) ){
```
Comment (mstaack) [inline app/Http/Controllers/AuthController.php]: if($verifyUser) is enough and more readable

File: app/Http/Controllers/AuthController.php [hunk 6]
```diff
@@ -52,26 +76,134 @@ public function login(Request $request)
...
+
+    public function refresh(Request $request)
+    {
+        if ( $token = Auth::refresh($request) ) {
+            return response([
+                'status' => 'success'
+            ])
+            ->header('Authorization', $token);
         }
+        else {
```
Comment (mstaack) [inline app/Http/Controllers/AuthController.php]: no return needed

File: app/User.php [hunk 7]
```diff
@@ -29,8 +29,14 @@ class User extends Model implements AuthenticatableContract, AuthorizableContrac
      */
     protected $hidden = [
         'password',
+        'refresh_token',
     ];
 
+    public function verifyUser()
```
Comment (mstaack) [inline app/User.php]: doc blocks

File: app/VerifyUser.php [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+class VerifyUser extends Model
+{
+    protected $guarded = [];
+    protected $primaryKey = 'user_id';
+    
+    public function user()
+    {
+        return $this->belongsTo('App\User', 'user_id');
+    }
+
```
Comment (mstaack) [inline app/VerifyUser.php]: spacing

**Review by mstaack — CHANGES_REQUESTED** [review]

**Comment by stefanvanherwijnen** [comment]
Sorry for the late reply.
Thanks for the feedback. I will try to correct it tomorrow.

File: app/Http/Controllers/AuthController.php (hunk 4 above)
Comment (stefanvanherwijnen) [reply app/Http/Controllers/AuthController.php]: Nevermind... I interpreted this the wrong way.

File: app/Http/Controllers/AuthController.php (hunk 6 above)
Comment (stefanvanherwijnen) [reply app/Http/Controllers/AuthController.php]: Without returning an error message, how could the user know that the refresh token is invalid and that he needs to login again?

Comment (mstaack) [reply app/Http/Controllers/AuthController.php]: sry meant no else needed

**Comment by mstaack** [comment]
@stefanvanherwijnen will finalize this pr today..  really love openapi 3.0!!! thanks

**Comment by mstaack** [comment]
@stefanvanherwijnen i wrote most of the stuff from scratch...see master branch.
regarding openapi & refresh token process i will rethink how i want to approach that.

thanks for your hint&ideas! sry that i didnt directly merge your stuff

cheers max

**Pull request closed without merging**